packages/react-router/lib/router/router.ts

```diff
@@ -4849,78 +4852,85 @@ async function loadLazyRouteModule(
   mapRouteProperties: MapRoutePropertiesFunction,
   manifest: RouteManifest
 ) {
+  let routeToUpdate = manifest[route.id];
+  invariant(routeToUpdate, "No route found in manifest");
+
   if (!route.lazy) {
     return;
   }
 
-  let lazyRoute = await route.lazy();
-
-  // If the lazy route function was executed and removed by another parallel
-  // call then we can return - first lazy() to finish wins because the return
-  // value of lazy is expected to be static
-  if (!route.lazy) {
+  // Check if we have a cached promise from a previous call
+  let cachedPromise = lazyRouteFunctionCache.get(routeToUpdate);
+  if (cachedPromise) {
+    await cachedPromise;
     return;
   }
 
-  let routeToUpdate = manifest[route.id];
-  invariant(routeToUpdate, "No route found in manifest");
-
-  // Update the route in place.  This should be safe because there's no way
-  // we could yet be sitting on this route as we can't get there without
-  // resolving lazy() first.
-  //
-  // This is different than the HMR "update" use-case where we may actively be
-  // on the route being updated.  The main concern boils down to "does this
-  // mutation affect any ongoing navigations or any current state.matches
-  // values?".  If not, it should be safe to update in place.
-  let routeUpdates: Record<string, any> = {};
-  for (let lazyRouteProperty in lazyRoute) {
-    let staticRouteValue =
-      routeToUpdate[lazyRouteProperty as keyof typeof routeToUpdate];
-
-    let isPropertyStaticallyDefined =
-      staticRouteValue !== undefined &&
-      // This property isn't static since it should always be updated based
-      // on the route updates
-      lazyRouteProperty !== "hasErrorBoundary";
+  let lazyRoutePromise = route.lazy().then((lazyRoute) => {
```

Comment:
I _think_ the reason to mix `await`/`.then` here is to ensure that once `await cachedPromise` resolves above, the route is guaranteed to have been updated.  It took me a bit to figure out why we were mixing them and also to see that there doesn't seem to be a great way to do that with just await without extracting to a new function.  

If that's correct, maybe worth adding a short comment above this line just to explain to future us 😂